docs/devel/kubectl-conventions.md

```diff
@@ -135,6 +133,151 @@ Updated: 8/27/2015
 * Use "TYPE" for the particular flavor of resource type accepted by kubectl, rather than "RESOURCE" or "KIND"
 * Use "NAME" for resource names
 
+## Command implementation conventions
+
+For every command there should be a `NewCmd<CommandName>` function that creates the command and returns a pointer to a `cobra.Command`, which can later be added to other parent commands to compose the structure tree. There should also be a `<CommandName>Options` struct with a variable to every flag and argument declared by the command (and any other variable required for the command to run). This makes tests and mocking easier. The struct ideally exposes three functions:
+
+* `Complete`: Completes the struct variables with values that may not be directly provided, for example, by flags pointers. Here you will usually take the `args` slice and set the values as appropriate variables, instantiate configs or clients, etc.
+* `Validate`: performs validation and returns errors.
+* `Run<CommandName>`: runs the actual command, taking as assumption that the struct is complete with all required values to run, and they are valid.
```

Comment:
I think #7311 is slightly different; that suggests moving all business logic out of command definitions, which sounds like a good idea. Inside the command we still need access to the `cobra.Command`, `cmdutil.Factory`, `cmdIn`, and `cmdOut`, to parse flag values, access generators/printer, etc. Current convention is to do that in `Run<CommandName>`, which can (and should) use a library to handle the business logic aside from parsing input and formatting output. e.g. `rollingupdate` [does this](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/rollingupdate.go#L284).